app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt

```diff
@@ -0,0 +1,1166 @@
+          _selectedHike.value = hike.value
+          successful = true
+        }
+
+        // Perform the callback without the lock, to avoid deadlocks and improve performance
+        if (successful) {
+          onSuccess()
+        } else {
+          onFailure()
+        }
+      }
+
+  private suspend fun unselectHikeAsync() =
+      _hikesMutex.withLock {
+        // Only emit null as a value if the selected hike was not already null
+        if (_selectedHikeId != null) {
+          _selectedHikeId = null
+          _selectedHike.value = null
+        }
+      }
+
+  /**
+   * This function performs a loading operation but DOES NOT set [_loading]. If calling it from
+   * another function, please set [_loading] to true before calling the function and to false once
+   * the call has ended.
+   *
+   * @return True if the operation is successful, false otherwise.
+   */
+  private suspend fun refreshSavedHikesCacheAsync(
+      onSuccess: () -> Unit = {},
+      onFailure: () -> Unit = {}
+  ): Boolean =
+      withContext(dispatcher) {
+        val savedHikes: List<SavedHike>
+        try {
+          // Load the saved hikes from the repository
+          savedHikes = savedHikesRepo.loadSavedHikes()
+        } catch (e: Exception) {
+          Log.e(LOG_TAG, "Error encountered while loading saved hikes", e)
+          onFailure()
+          return@withContext false
+        }
+
+        // Wait for the lock to avoid concurrent modifications
+        _hikesMutex.withLock { updateSavedHikesCache(savedHikes) }
+
+        onSuccess()
+        return@withContext true
+      }
+
+  /**
+   * Helper function for [refreshSavedHikesCacheAsync].
+   *
+   * Takes a list of saved hikes and updates the local cache of saved hikes with that list.
+   *
+   * It is the caller's responsibility to ensure the saved hikes list is up-to-date and to acquire
+   * [_hikesMutex] before calling this function.
+   *
+   * @param newList The list of saved hikes to update the cache with.
+   */
+  private fun updateSavedHikesCache(newList: List<SavedHike>) {
+    // Clear the current saved hikes register
+    _savedHikesMap.clear()
+
+    // Build the new saved hikes register
+    newList.forEach { savedHike -> _savedHikesMap[savedHike.id] = savedHike }
+
+    // Update the map of hikes to reflect the changes
+    val toRemove: MutableSet<String> = mutableSetOf()
+    _hikeFlowsMap.values.forEach { hikeFlow ->
+      val hike = hikeFlow.value
+      val savedHike = _savedHikesMap[hike.id]
+      // If the list currently only contains saved hikes and this hike was unsaved, remove it
+      if (_loadedHikesType == LoadedHikes.FromSaved && savedHike == null) {
+        toRemove.add(hike.id)
+      }
+      // Otherwise, check if a change actually happened to avoid unnecessary updates
+      else {
+        val (changeNeeded, updated) = hikeNeedsSavedStatusUpdate(hike, savedHike)
+        if (changeNeeded) {
+          hikeFlow.value = updated
+        }
```

Comment:
See my previous answer for why I do not simply update it.

Indeed, updating without needing changes would simply set the hike to the same value as before, but it might trigger UI updates anyway and be less efficient if all hikes are updated at each refresh.

Reply:
Thanks for the answer!